## Fixing Model Overfitting Using Lasso Regression
The goal of this exercise is to teach you how to identify when your model starts overfitting, and to use lasso regression to fix overfitting in your model.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso
from sklearn.metrics import mean_squared_error
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler, PolynomialFeatures

In [2]:
_df = pd.read_csv('https://raw.githubusercontent.com/'\
                 'PacktWorkshops/The-Data-Science-Workshop/'\
                 'master/Chapter07/Dataset/ccpp.csv')
_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9568 entries, 0 to 9567
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AT      9568 non-null   float64
 1   V       9568 non-null   float64
 2   AP      9568 non-null   float64
 3   RH      9568 non-null   float64
 4   PE      9568 non-null   float64
dtypes: float64(5)
memory usage: 373.9 KB


- Temperature (T) in the range 1.81°C and 37.11°C,
- Ambient Pressure (AP) in the range 992.89-1033.30 milibar,
- Relative Humidity (RH) in the range 25.56% to 100.16%
- Exhaust Vacuum (V) in teh range 25.36-81.56 cm Hg
- Net hourly electrical energy output (PE) 420.26-495.76 MW

In [3]:
# features and labels
X = _df.drop(['PE'], axis=1).values
y = _df['PE'].values

In [4]:
# split data into training and evaluation sets
train_X, eval_X, train_y, eval_y = train_test_split(X, y, train_size=0.8, random_state=0)
#val_X, test_X, val_y, test_y = train_test_split(eval_X, eval_y, random_state=0)

In [5]:
# instantiate LinearRegression
lr_model_1 = LinearRegression()

#fit model
lr_model_1.fit(train_X, train_y)

LinearRegression()

In [6]:
# make predictions on the evaluation dataset
lr_model_1_preds = lr_model_1.predict(eval_X)

In [7]:
# R2 of the model
print('lr_model_1 Score: {}'.format(lr_model_1.score(eval_X, eval_y)))

lr_model_1 Score: 0.9325315554761303


In [8]:
# MSE
print('lr_model_1 MSE: {}'.format(mean_squared_error(eval_y, lr_model_1_preds)))

lr_model_1 MSE: 19.733699303497637


The first model was trained on four features. You will now train a new model on four cubed features.

In [9]:
# create a list of tuples to serve as a pipeline
steps = [('scaler', MinMaxScaler()), 
         ('poly', PolynomialFeatures(degree=3)), 
        ('lr', LinearRegression())]

In this step, you create a list with three tuples. The first tuple represents a scaling operation that makes use of MinMaxScaler. The second tuple represents a feature engineering step and makes use of PolynomialFeatures. The third tuple represents a LinearRegression model.

The first element of the tuple represents the name of the step, while the second element represents the class that performs a transformation or an estimator.

In [10]:
# create an instance of a pipeline
lr_model_2 = Pipeline(steps)

In [11]:
# train pipeline instance
lr_model_2.fit(train_X, train_y)

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('poly', PolynomialFeatures(degree=3)),
                ('lr', LinearRegression())])

The pipeline implements a .fit() method, which is also implemented in all instances of transformers and estimators. The .fit() method causes .fit_transform() to be called on transformers, and causes .fit() to be called on estimators.

In [12]:
# R2 of model 2
print('lr_model_2 R2 score: {}'.format(lr_model_2.score(eval_X, eval_y)))

lr_model_2 R2 score: 0.9443678654045206


In [13]:
# preds of model 2
lr_model_2_preds = lr_model_2.predict(eval_X)

In [14]:
# MSE of model 2
print('lr_model_2 MSE: {}'.format(mean_squared_error(eval_y, lr_model_2_preds)))

lr_model_2 MSE: 16.27172263220768


You can see from the output that the MSE of the second model is 16.27. This is less than the MSE of the first model, which is 19.73. You can safely conclude that the second model is better than the first.

In [15]:
# inspect model coefficients (weights)
print(lr_model_2[-1].coef_)

[ 7.72661789e-14 -1.77278028e+02 -4.60337188e+01 -1.60520675e+02
 -1.23076123e+02  6.23358210e+00  8.19655844e+00  1.45478576e+02
  1.88658651e+02  2.43740192e+01  1.80553150e+02 -1.08058561e+02
  1.09713294e+02  1.79121906e+02  1.06460596e+02  2.67290613e+01
  7.79833654e+01  3.69241324e+01 -1.13863997e+02 -1.42673215e+02
 -9.69606773e+01  1.90706809e+02 -5.56429546e+01 -1.32595225e+02
 -9.41682917e+01  9.40112729e+01 -1.18732510e+02 -7.64871610e+01
 -4.18714081e+01  6.36772260e+01  4.42340977e+01 -3.81114691e+01
 -4.71547759e+01 -9.16797074e+01 -2.52346805e+01]


In this step, you will note that lr_model_2 is a pipeline. The final object in this pipeline is the model, so you make use of list addressing to access this by setting the index of the list element to -1.

Once you have the model, which is the final element in the pipeline, you make use of .coef_ to get the model coefficients

In [16]:
# check for the number of coefficients in this model
print(len(lr_model_2[-1].coef_))

35


In [17]:
# create a steps list with PolynomialFeatures of degree 10
steps = [('scaler', MinMaxScaler()), 
        ('poly', PolynomialFeatures(degree=10)), 
        ('lr', LinearRegression())]

In [18]:
# create model 3
lr_model_3 = Pipeline(steps)

In [19]:
# fit model 3
lr_model_3.fit(train_X, train_y)

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('poly', PolynomialFeatures(degree=10)),
                ('lr', LinearRegression())])

In [20]:
# R2 of model 3
print('lr_model_3 R2 score: {}'.format(lr_model_3.score(eval_X, eval_y)))

lr_model_3 R2 score: 0.5683445811859165


You can see from the preceding figure that the R2 score is now 0.56. The previous model had an R2 score of 0.944. This model has an R2 score that is considerably worse than the one of the previous model, lr_model_2. This happens when your model is overfitting.

In [21]:
# preds for model 3
lr_model_3_preds = lr_model_3.predict(eval_X)

In [22]:
# MSE of model 3
print('lr_model_3_preds MSE: {}'.format(mean_squared_error(eval_y, lr_model_3_preds)))

lr_model_3_preds MSE: 126.25395913179554


In [23]:
# print coefficients of model 3
print(len(lr_model_3[-1].coef_))

1001


In [24]:
# inspect first 35 weights to get a sense of the individual magnitueds
print(lr_model_3[-1].coef_[:35])

[ 3.92505101e+05 -6.90885527e+07 -4.12732195e+07  2.27924135e+07
 -4.76789946e+07  2.96662372e+08  2.73270121e+08  1.07845355e+08
  3.73718730e+08  8.79697148e+07 -2.35335367e+07  2.46253911e+08
 -2.61103329e+08  1.86100397e+07  1.41131427e+08 -6.53882597e+08
 -8.90637240e+08 -1.06074229e+09 -1.29264008e+09 -4.28439276e+08
  5.31443921e+07 -1.30409864e+09  4.41022600e+08 -8.86227463e+08
 -8.78158963e+08 -1.97159667e+06 -5.39373932e+08 -3.68353344e+08
  9.82102192e+08 -2.76731667e+08 -6.28828639e+08  8.14253722e+08
  5.43202144e+08 -2.03046371e+08 -2.42929104e+08]


In [25]:
# pipeline with lasso
steps = [('scaler', MinMaxScaler()), 
        ('poly', PolynomialFeatures()), 
        ('lr', Lasso(alpha=0.01))]

You create a list of steps for the pipeline you will create. Note that the third step in this list is an instance of lasso. The parameter called alpha in the call to Lasso() is the regularization parameter. You can play around with any values from 0 to 1 to see how it affects the performance of the model that you train.

In [26]:
# create instance of pipeline
lasso_model = Pipeline(steps)

In [27]:
# fit pipeline on training data
lasso_model.fit(train_X, train_y)

Pipeline(steps=[('scaler', MinMaxScaler()), ('poly', PolynomialFeatures()),
                ('lr', Lasso(alpha=0.01))])

In [28]:
# lasso R2
print('lasso_model R2 score: {}'.format(lasso_model.score(eval_X, eval_y)))

lasso_model R2 score: 0.9381690882453707


In [29]:
# preds for lasso model
lasso_preds = lasso_model.predict(eval_X)

In [30]:
# lasso MSE
print('lasso_model MSE: {}'.format(mean_squared_error(eval_y, lasso_preds)))

lasso_model MSE: 18.08478954624889


In [32]:
# number of lasso weights
print(len(lasso_model[-1].coef_))

15


In [35]:
print(lasso_model[-1].coef_[:35])

[  0.         -76.85674783  -7.21377479   2.81063224  -0.
  13.32449178   0.           0.          -6.0078807    0.
   0.         -11.17782095  -0.           0.          -0.81595656]
